### DISCUSSION TITLE
Creating custom SignIn pages with Authjs-Sveltekit

### DISCUSSION THREAD

**Discussion by OwenRestock** [nextauthjs/next-auth#7287] [Help]
### Question 💬

Hi
I would like to create a custom login / signIn page for my app, but after do some research I can't found any example using sveltekit, instead I found this tutorial using [nextjs](https://www.youtube.com/watch?v=hADeo48SATU) I've tried those tutorials but the signIn method within the custom signIn page doesn't seems to work, it only refreshing the page and doesn't throw any error (I already set the debug props to true, and add the custom **pages.signIn** route within the **SvelteKitAuth** declaration). 

Is the signIn method behavior considered as a bug or creating custom signIn pages is not supported yet for the sveltekit adapter?

Thanks for your response, appreciate it!

### How to reproduce ☕️

Currently I'm using @auth/core: 0.5.1 and @auth/sveltekit: 0.3.0
I'm using email-password credential.

        <script>
        import { signIn } from '@auth/sveltekit/client';
        
        const handleLogin = async () => {
                console.log({ email, password });
                await signIn('email-password', {
                    email: 'email@email.com',
                    password: 'password',
                    redirect: false,
                    callbackUrl: '/',
                });
                return
            };
        </script>

        <div>
            <h1 class="notSignedInText">You are not signed in</h1>
            <br />
            <label for="email">Email:</label>
            <input type="email" bind:value={email} />
            <br />
            <br />
            <label for="password">Password:</label>
            <input type="password" bind:value={password} />
            <br />
            <br />
            <button
                class="bg-emerald-400 px-4 py-2"
                type="button"
                on:click={handleLogin}
            >
                Login with Email & Password
            </button>
        </div>

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by ziedHamdi** [comment]
I'm also searching how to do that, I can't find how to obtain the list of providers from the `SvelteKitAuth `instance I created. I couldn't even find where the default implementation is, seems as if it is still in react

**Reply by tdack** [reply]
Fetch them from `/auth/providers` in either a `page.js` or a `page.server.js` and then look at the NextJS example at https://authjs.dev/guides/basics/pages or https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/lib/pages/signin.tsx to see how the providers are looped over to generate the login buttons.

**Comment by ziedHamdi** [comment]
This is what I did so far:

**hooks.server.js** //this one works, I just added the pages option to redirect to my page
```
import { SvelteKitAuth } from "@auth/sveltekit";
import LinkedIn from '@auth/core/providers/linkedin';
import Google from "@auth/core/providers/google";

import { MongoDBAdapter } from "@next-auth/mongodb-adapter"
import conn from './lib/server/db/mongoClient'

import { env } from '$env/dynamic/private'
import { redirect } from '@sveltejs/kit';

const { LINKED_IN_CLIENT_ID, LINKED_IN_CLIENT_SECRET, GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET } = env

export const handle = async (options) => {
	//Auth.js
	const auth = SvelteKitAuth({
		adapter: MongoDBAdapter(conn),
		secret:env.AUTH_SECRET,
		providers: [
			// LinkedIn({ clientId: LINKED_IN_CLIENT_ID, clientSecret: LINKED_IN_CLIENT_SECRET }),
			Google({ clientId: GOOGLE_CLIENT_ID, clientSecret: GOOGLE_CLIENT_SECRET }),
		],
		callbacks: {
			async session(session, user) {
				return session;
			},
		},
		pages: {
			signIn: '/auth/signin',
			// signOut: '/auth/signout',
			// error: '/auth/error', // Error code passed in query string as ?error=
			// verifyRequest: '/auth/verify-request', // (used for check email message)
			// newUser: '/auth/new-user' // New users will be directed here on first sign in (leave the property out if not of interest)
		}
	})

	//redirect unauthentified users to homepage
	// if (!session?.user && event.url.href.includes('authorized')) {
	// 	throw redirect(302, '/');
	// }

	const response = await auth(options);
	const { resolve, event } = options


	// Apply CORS header for API routes
	if (event.url.pathname.startsWith('/authorized/cv/submit')) {
		// Required for CORS to work
		if(event.request.method === 'OPTIONS') {
			return new Response(null, {
				headers: {
					'Access-Control-Allow-Methods': 'POST, GET, OPTIONS, DELETE',
					'Access-Control-Allow-Origin': '*',
				}
			});
		}
		response.headers.append('Access-Control-Allow-Origin', `*`);
	}
	return response;
};
```

**src/routes/auth/signin/+page.svelte**
```
<script>
	import { signIn, getSession, csrfToken } from "@auth/sveltekit/client";

	export let data;
</script>

<div>
		{#each data.providers as provider, index(provider.name)}
			<div key={provider.name}>
				<button on:click={() => signIn(provider.id)}>
					Sign in with {provider.name}
				</button>
			</div>
			{/each}
</div>
```

**src/routes/auth/signin/+page.server.js** //putting the providers manually
```
import Google from "@auth/core/providers/google";

export async function load(event) {
	const providers = [Google]
	return {providers:JSON.parse(JSON.stringify(providers))}
}
```

This sends me to a page with this url http://localhost:5173/auth/signin?callbackUrl=http%3A%2F%2Flocalhost%3A5173%2Fauthorized%2Fcv%2F647a29e46ba81ab341e4bf4a
http://localhost:5173/auth/signin?callbackUrl=http%3A%2F%2Flocalhost%3A5173%2Fauthorized%2Fcv%2F647a29e46ba81ab341e4bf4a

And it complains of too many redirects

**Comment by H-G-11** [comment]
I also have the too many redirects when I specify 
```
pages: {
  signIn: "/auth/signin",
  },
```

When I specify 
```
pages: {
  signIn: "/signin",
  },
```

I don't have too many redirects, but the authentication seems broken (nothing happens).

I prefer not investigating too much time in that right now, I know that this doesn't help much. I will probably have a look again in some time.

Thank you very much

**Reply by bbauti** [reply]
Hey! did you find an solution?

**Reply by ziedHamdi** [reply]
I integrated Lucia to my project, it took me a day to have Google and email signup, then a day to handle mail verification and password changing. Definitely worth it.

**Reply by joelriveradev-endava** [reply]
I figured out why the "too many redirects" problem occurs. It's because of a mismatch between the key and the expected route assigned to it. It should be: 

```TypeScript 
pages: {
   signIn: '/auth/signIn',
   signOut: '/auth/signOut'
}
```
Notice how signIn is with an uppercase "I", not with a lowercase. When I realized this, it fixed my problem and I now have a custom sign in page. For the providers, I just send back an array of providers from the server, along with some additional metadata to create the ui.

```TypeScript
export const load: PageServerLoad = async ({ locals, url }) => {
  const session = await locals.getSession()
  const callbackUrl = new URL(url).searchParams.get('originalUrl') ?? '/'

  return {
    session,
    providers: [
      { id: 'google', label: 'Google', callbackUrl },
      { id: 'facebook', label: 'Facebook', callbackUrl },
      { id: 'github', label: 'GitHub', callbackUrl },
    ]
  }
}
```

Hope this helps.

**Reply by jasongitmail** [reply]
@ziedHamdi Lucia looks like so many little pieces to put together. Isn't AuthJS easier? I was able to get AuthJS OAuth working in an hour or two.

**Reply by ziedHamdi** [reply]
Yes me too, but then, I had to find how to create my own login page, struggled in it, searched for svelte specific documentation and ended up in front of the source code of the lib. Which is pretty bad news, as you have to understand the way things are organized, with many surprises down the road. 
So at the moment I knew I cannot estimate the time it will take me to get through, I searched for alternatives, and Lucia did the job.

**Reply by jasongitmail** [reply]
Got it. Thanks for sharing your experience.

I wish a higher level of abstraction existed on top of Lucia, or AuthJS docs were a bit more polished. I'm [trying to help a bit with the AuthJS docs as I use AuthJS](https://github.com/nextauthjs/next-auth/pull/8833/files) but let's see how it goes. I'll keep an open mind to Lucia.

**Reply by ziedHamdi** [reply]
Happy to talk @jasongitmail. It's been a while since I implemented Lucia, I'm the kind of people who forget fast :-). So I can't tell you what I liked and what I appreciated. I remember though that for AuthJs there was no high level architecture document to give directions on where to search for a given feature. It was even more true for Svelte

**Reply by ziedHamdi** [reply]
@jasongitmail I'm writing a project to allow low-entry ticket for pay-as-you-go integration. I'd be very happy to have contributors if you are interested. It's still early stage, but I'm advancing pretty fast on it. Here's an article I wrote about how it does what it does: https://dev.to/zhamdi/pay-as-you-go-a-flexible-approach-to-managing-multiple-offers-simultaneously-for-a-user-2a1m

I hope it's not intrusive to talk about it here, I feel like we are a community of open source lovers who help each other, but I'm not sure if I understand all behavior codes.

**Reply by jasongitmail** [reply]
Cool. It's definitely worth sharing your project to see if any devs may have aligned interests and want to collaborate.  

Are there any open source billing or SaaS-related projects that have discussions groups where you could share it and ask for feedback? Or I've found some success sharing code on Github and then people eventually come along to ask questions or add issues, and they have similar goals and are a good fit to collaborate with. Might find devs who are a good fit there. I don't have expertise in payments or need in that area currently, but likely others do.

**Reply by ziedHamdi** [reply]
Thanks for the tips, actually I [posted a question](https://www.reddit.com/r/sveltejs/comments/16tfvcp/pay_as_you_go_js_library/) searching for an open source lib on reddit, and from the answers it seems there are not any yet. So as I have to implement it for my project anyway, I decided to build it in public. So I'm still at the beginning of the adventure.

Happy we talked @jasongitmail, take care

**Reply by jasongitmail** [reply]
Likewise! Best of luck with your project.

**Comment by hazlanmshabli** [comment]
https://medium.com/@uriser/authentication-in-sveltekit-with-auth-js-7ff505d584c4 this looks promising. I am trying it out

**Comment by coryvirok** [comment]
Apparently it is **not possible** to host a custom page on `/auth/signin`, `/auth/error`, etc. using SvelteKit due to this code: https://github.com/nextauthjs/next-auth/blob/39dd3b92de194c1a835f2d87631f4deb9d9fdf65/packages/frameworks-sveltekit/src/lib/index.ts#L380-L385

If you create `SvelteKitAuth` with `pages: { signIn: '/auth/signin' }` and you add a route for a custom sign in experience at `src/routes/auth/signin/+page.svelte` you will see `ERR_TOO_MANY_REDIRECTS`.

Is there a way to do this that I'm missing?

**Comment by ohack38** [comment]
Since this might still be an unclear part of the auth flow, I thought i'd share a working solution. As @coryvirok mentioned using those default routes will not work so you need to create your own custom routes for these pages and link them inside your SvelteKitAuth config using the pages option.
``` javascript
pages: {
      signIn: "/signin",
      signOut: "/signout",
      verifyRequest: "/auth/verify",
      newUser: "/auth/new",
      error: "/auth/errors"
    },
```
Make sure you have those corresponding routes with regular ``` +page.svelte ``` files

**Comment by silvestrevivo** [comment]
This is working for me:

```
<SignIn provider="credentials">
  <span slot="submitButton">Sign In with Credentials</span>
  <span slot="credentials">
    <input type="text" name="username" value="hacky" />
    <input type="text" name="password" value="mypassword" />
    <input type="text" name="email" value="hacky@hacky.com" />
    <input type="hidden" name="redirectTo" value="/dashboard" />
  </span>
</SignIn>
```
